pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -688,23 +688,27 @@ func (m *ManagerImpl) takeByTopology(resource string, available sets.String, aff
 	// available device does not have any NUMA Nodes associated with it, add it
 	// to a list of NUMA Nodes for the fake NUMANode -1.
 	perNodeDevices := make(map[int]sets.String)
+	nodeWithoutTopology := -1
+	perNodeDevices[nodeWithoutTopology] = sets.NewString()
 	for d := range available {
-		var nodes []int
+		nodeAdded := false
 		if m.allDevices[resource][d].Topology != nil {
 			for _, node := range m.allDevices[resource][d].Topology.Nodes {
-				nodes = append(nodes, int(node.ID))
+				nodeAdded = true
+				nodeID := int(node.ID)
+				if _, ok := perNodeDevices[nodeID]; !ok {
+					perNodeDevices[nodeID] = sets.NewString()
+				}
+				perNodeDevices[nodeID].Insert(d)
 			}
 		}
-		if len(nodes) == 0 {
-			nodes = []int{-1}
-		}
-		for _, node := range nodes {
-			if _, ok := perNodeDevices[node]; !ok {
-				perNodeDevices[node] = sets.NewString()
-			}
-			perNodeDevices[node].Insert(d)
+		if !nodeAdded {
+			perNodeDevices[nodeWithoutTopology].Insert(d)
 		}
 	}
+	if perNodeDevices[nodeWithoutTopology].Len() == 0 {
+		delete(perNodeDevices, -1)
+	}
```

Comment:
I agree with this change in general. However, for readability, I would prefer the following.
It makes the two cases symmetrical and easier to follow and verify its correctness.

```
        nodeWithoutTopology := -1
        for d := range available {
                if m.allDevices[resource][d].Topology == nil || len(m.allDevices[resource][d].Topology.Nodes) == 0 {
                        if _, ok := perNodeDevices[nodeWithoutTopology]; !ok {
                                perNodeDevices[nodeWithoutTopology] = sets.NewString()
                        }
                        perNodeDevices[nodeWithoutTopology].Insert(d)
                        continue
                }

                for _, node := range m.allDevices[resource][d].Topology.Nodes {
                        if _, ok := perNodeDevices[int(node.ID)]; !ok {
                                perNodeDevices[int(node.ID)] = sets.NewString()
                        }
                        perNodeDevices[int(node.ID)].Insert(d)
                }
        }
```